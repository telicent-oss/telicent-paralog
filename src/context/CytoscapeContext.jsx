import { isEmpty } from "lodash";
import React, { createContext, useCallback, useRef } from "react";

import { useLocalStorage } from "hooks";
import { getUniqueElements } from "utils";

export const CytoscapeContext = createContext();

export const CytoscapeProvider = ({ children }) => {
  const cyRef = useRef({});
  const [layout, setLayout] = useLocalStorage("graphLayout", "cola");

  const moveTo = ({ areaSelect, cachedElements, selectedElements }) => {
    if (!cyRef.current) return;
    const padding = 20;
    const selected = areaSelect
      ? selectedElements
      : getUniqueElements([...cachedElements, ...selectedElements]);

    const elements = cyRef.current.elements().filter((element) => {
      return selected.some((selectedElement) => {
        const data = element.data("element");
        return selectedElement.uri === data.uri;
      });
    });

    fit(elements, padding);
  };

  const fit = (elements, padding) => {
    if (!cyRef.current) return;
    cyRef.current.fit(elements, padding);
  };

  const updateLayout = (layout) => {
    setLayout(layout);
  };

  const resize = () => {
    if (isEmpty(cyRef?.current)) return;
    cyRef.current.resize();
  };

  const runLayout = useCallback(() => {
    if (isEmpty(cyRef?.current) || cyRef.current?.destroyed()) return;
    const cylayout = cyRef.current.layout({ name: layout });
    cylayout.run();
  }, [cyRef, layout]);

  return (
    <CytoscapeContext.Provider
      value={{ cyRef, layout, fit, moveTo, resize, runLayout, updateLayout }}
    >
      {children}
    </CytoscapeContext.Provider>
  );
};
