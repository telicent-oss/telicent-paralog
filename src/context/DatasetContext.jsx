import { createContext, useState } from "react";

const DatasetContext = createContext();

const DatasetProvider = ({ children }) => {
  const [selectedFloodAreas, setSelectedFloodAreas] = useState([]);
  const [expandedFloodAreas, setExpandedFloodAreas] = useState([]);
  const [selectedAssets, setSelectedAssets] = useState([]);
  const [expandedAssets, setExpandedAssets] = useState([]);

  const addSelectedFloodAreas = (polygonUris) => {
    setSelectedFloodAreas(polygonUris);
  };

  const addSelectedAssets = (assets) => {
    console.log({ assets });
    setSelectedAssets(assets);
  };

  const removeSelectedAsset = (selectedAsset) => {
    setSelectedAssets((assets) => assets.filter((asset) => asset !== selectedAsset));
  };

  const toggleExpandedAssets = (selectedGroup) => {
    const index = expandedAssets.findIndex((ontologyGroup) => ontologyGroup === selectedGroup);
    if (index === -1) {
      setExpandedAssets([...expandedAssets, selectedGroup]);
      return;
    }
    const filteredGroups = expandedAssets.filter(
      (ontologyGroup) => ontologyGroup !== selectedGroup
    );
    setExpandedAssets(filteredGroups);
  };

  const updateExpandedFloodAreas = (groups) => {
    setExpandedFloodAreas(groups);
  };

  return (
    <DatasetContext.Provider
      value={{
        expandedAssets,
        expandedFloodAreas,
        selectedFloodAreas,
        selectedAssets,
        addSelectedAssets,
        addSelectedFloodAreas,
        removeSelectedAsset,
        toggleExpandedAssets,
        updateExpandedFloodAreas,
      }}
    >
      {children}
    </DatasetContext.Provider>
  );
};

export { DatasetContext, DatasetProvider };
