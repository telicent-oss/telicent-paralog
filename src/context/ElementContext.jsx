import React, { useCallback, useReducer, useState } from "react";

import elementsReducer, {
  ADD_ASSETS,
  ADD_DEPENDENCIES,
  AREA_SELECTION,
  CLEAR_SELECTED,
  DISMISS_ERROR,
  FILTER_SELECTED_ELEMENTS,
  INITIAL_STATE,
  MULTI_SELECT_ELEMENTS,
  REMOVE_ELEMENTS_BY_TYPE,
  RESET,
  SELECT_ELEMENT,
  UPDATE_ERRORS,
} from "./elements-reducer";

export const ElementsContext = React.createContext();

export const ElementsProvider = ({ children }) => {
  const [state, dispatch] = useReducer(elementsReducer, INITIAL_STATE);
  const [selectedTimeline, setSelectedTimeline] = useState(null);

  const onFloodTimelineSelect = (selected) => {
    if (selected !== selectedTimeline) {
      setSelectedTimeline(selected);
    }
    return;
  };

  const closeTimelinePanel = () => {
    setSelectedTimeline(null);
  };

  const {
    assets,
    dependencies,
    errors,
    selectedElements,
    maxAssetCriticality,
    maxAssetTotalCxns,
    assetCriticalityColorScale,
    cxnCriticalityColorScale,
    totalCxnsColorScale,
  } = state;

  const filterSelectedElements = useCallback(() => {
    dispatch({ type: FILTER_SELECTED_ELEMENTS });
  }, []);

  const addElements = useCallback(
    (assets, dependencies) => {
      if (!Array.isArray(assets) || !Array.isArray(dependencies)) return;
      dispatch({ type: ADD_ASSETS, assets });
      dispatch({ type: ADD_DEPENDENCIES, dependencies });
      filterSelectedElements();
    },
    [filterSelectedElements]
  );

  const removeElementsByType = useCallback(
    (typeUri) => {
      if (!typeUri) return;
      dispatch({ type: REMOVE_ELEMENTS_BY_TYPE, typeUri });
      filterSelectedElements();
    },
    [filterSelectedElements]
  );

  const reset = useCallback(() => {
    dispatch({ type: RESET });
  }, []);

  const onElementClick = useCallback((multiSelect, selectedElements) => {
    if (multiSelect) {
      dispatch({ type: MULTI_SELECT_ELEMENTS, selectedElements });
      return;
    }
    dispatch({ type: SELECT_ELEMENT, selectedElements });
  }, []);

  const onAreaSelect = useCallback((selectedElements) => {
    if (!Array.isArray(selectedElements)) return;
    dispatch({ type: AREA_SELECTION, selectedElements });
  }, []);

  const updateErrorNotifications = useCallback((msg) => {
    dispatch({ type: UPDATE_ERRORS, error: msg });
  }, []);

  const dismissErrorNotification = (error) => {
    dispatch({ type: DISMISS_ERROR, error });
  };

  const clearSelectedElements = useCallback(() => {
    dispatch({ type: CLEAR_SELECTED });
  }, []);

  return (
    <ElementsContext.Provider
      value={{
        assets,
        dependencies,
        errors,
        selectedElements,
        maxAssetCriticality,
        maxAssetTotalCxns,
        assetCriticalityColorScale,
        cxnCriticalityColorScale,
        totalCxnsColorScale,
        addElements,
        clearSelectedElements,
        dismissErrorNotification,
        filterSelectedElements,
        onAreaSelect,
        onElementClick,
        reset,
        removeElementsByType,
        updateErrorNotifications,
        selectedTimeline,
        closeTimelinePanel,
        onFloodTimelineSelect,
      }}
    >
      {children}
    </ElementsContext.Provider>
  );
};
