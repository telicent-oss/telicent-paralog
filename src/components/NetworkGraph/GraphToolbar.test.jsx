import { render, screen, within } from "@testing-library/react";
import { vi } from "vitest";
import userEvent from "@testing-library/user-event";
import React, { useContext } from "react";
import { CytoscapeContext, CytoscapeProvider } from "../../context";
import Toolbar from "./GraphToolbar";

const user = userEvent.setup();

const ToolbarTestComponent = ({ cyRef }) => {
  const { layout, updateLayout } = useContext(CytoscapeContext);
  return (
    <Toolbar cyRef={cyRef} graphLayout={layout} setGraphLayout={updateLayout} />
  );
};

describe("GraphToolbar component", () => {
  test("renders all layout options", async () => {
    render(<ToolbarTestComponent />, { wrapper: CytoscapeProvider });
    const layoutBtn = screen.getByRole("button", { name: /layout/i });
    await user.hover(layoutBtn);
    expect(screen.getByText("Layout")).toBeVisible();
    expect(screen.queryByTestId("secondary-menu")).not.toBeInTheDocument();

    await user.click(layoutBtn);

    const secondaryMenuItems = within(
      screen.getByTestId("secondary-menu"),
    ).getAllByRole("listitem");
    expect(secondaryMenuItems).toHaveLength(6);
    expect(secondaryMenuItems).toMatchSnapshot("layout options");
  });

  test("renders Cola as the default graph layout", async () => {
    render(<ToolbarTestComponent />, { wrapper: CytoscapeProvider });
    await user.click(screen.getByRole("button", { name: /layout/i }));

    const secondaryMenuItems = within(
      screen.getByTestId("secondary-menu"),
    ).getAllByRole("listitem");
    expect(
      within(secondaryMenuItems[0]).getByRole("button", { name: "Cola" }),
    ).toHaveClass("bg-black-500");
  });

  test("renders updated graph layout", async () => {
    render(<ToolbarTestComponent />, { wrapper: CytoscapeProvider });
    await user.click(screen.getByRole("button", { name: /layout/i }));

    const secondaryMenuItems = within(
      screen.getByTestId("secondary-menu"),
    ).getAllByRole("listitem");
    await user.click(
      within(secondaryMenuItems[1]).getByRole("button", { name: "Grid" }),
    );

    expect(
      within(secondaryMenuItems[1]).getByRole("button", { name: "Grid" }),
    ).toHaveClass("bg-black-500");
    expect(
      within(secondaryMenuItems[0]).getByRole("button", { name: "Cola" }),
    ).not.toHaveClass("bg-black-500");
  });

  test("pans and zooms the graph to fit", async () => {
    const mockFit = vi.fn();
    const cy = { current: { fit: mockFit } };
    render(<ToolbarTestComponent cyRef={cy} />, { wrapper: CytoscapeProvider });

    const fitBtn = screen.getByRole("button", { name: /fit/i });
    await user.hover(fitBtn);
    expect(screen.getByText("Fit")).toBeVisible();

    await user.click(fitBtn);
    expect(mockFit).toHaveBeenCalledTimes(1);
  });

  test("pans the graph to the centre", async () => {
    const mockCenter = vi.fn();
    const cy = { current: { center: mockCenter } };
    render(<ToolbarTestComponent cyRef={cy} />, { wrapper: CytoscapeProvider });

    const centerBtn = screen.getByRole("button", { name: /center/i });
    await user.hover(centerBtn);
    expect(screen.getByText("Center")).toBeVisible();

    await user.click(centerBtn);
    expect(mockCenter).toHaveBeenCalledTimes(1);
  });

  test("exports current graph view as png", async () => {
    const mockPng = vi.fn();
    const mockRevokeObjectURL = vi.fn();
    const href =
      "blob:http://localhost:3001/eb1bb2f8-9e7d-4157-9418-cafe78aba65a";
    const cy = { current: { png: mockPng } };
    global.window.URL.createObjectURL = vi.fn().mockReturnValue(href);
    global.window.URL.revokeObjectURL = mockRevokeObjectURL;
    HTMLAnchorElement.prototype.click = vi.fn();
    render(<ToolbarTestComponent cyRef={cy} />, { wrapper: CytoscapeProvider });

    const exportBtn = screen.getByRole("button", { name: /export/i });
    await user.hover(exportBtn);
    expect(screen.getByText("Export")).toBeVisible();

    await user.click(exportBtn);
    expect(mockPng).toHaveBeenCalledTimes(1);
    expect(mockRevokeObjectURL).toHaveBeenCalledTimes(1);
    expect(mockRevokeObjectURL).toHaveBeenCalledWith(href);
  });
});
