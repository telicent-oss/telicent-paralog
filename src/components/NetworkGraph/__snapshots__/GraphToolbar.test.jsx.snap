// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GraphToolbar component > renders all layout options > layout options 1`] = `
[
  <li
    class="w-full"
  >
    <button
      class="hover:bg-black-400 px-2 rounded-md w-full h-full whitespace-nowrap bg-black-500"
    >
      Cola
    </button>
  </li>,
  <li
    class="w-full"
  >
    <button
      class="hover:bg-black-400 px-2 rounded-md w-full h-full whitespace-nowrap"
    >
      Grid
    </button>
  </li>,
  <li
    class="w-full"
  >
    <button
      class="hover:bg-black-400 px-2 rounded-md w-full h-full whitespace-nowrap"
    >
      Random
    </button>
  </li>,
  <li
    class="w-full"
  >
    <button
      class="hover:bg-black-400 px-2 rounded-md w-full h-full whitespace-nowrap"
    >
      Breadth First
    </button>
  </li>,
  <li
    class="w-full"
  >
    <button
      class="hover:bg-black-400 px-2 rounded-md w-full h-full whitespace-nowrap"
    >
      AVSDF
    </button>
  </li>,
  <li
    class="w-full"
  >
    <button
      class="hover:bg-black-400 px-2 rounded-md w-full h-full whitespace-nowrap"
    >
      Dagre
    </button>
  </li>,
]
`;
